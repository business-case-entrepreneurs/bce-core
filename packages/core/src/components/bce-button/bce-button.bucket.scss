@import './bce-button.scss';

:host {
  @include bce-color-default(primary);
  @include elevate(0);
  border-radius: 0;

  button:after {
    content: '';
    position: absolute;
    height: 4px;
    left: 0;
    bottom: 0;
    right: 0;
    background-color: #000;
    opacity: 0.25;
    border-radius: 0 0 0 0;
  }
}

:host([focus]) {
  @include elevate(0);
}

:host([icon]) button {
  span {
    padding-left: 36px;
  }

  bce-icon {
    color: bce-color();
    background-color: bce-color(null, 'icon');
  }

  &[data-icon-only] bce-icon {
    background-color: bce-color();
    color: bce-color-t();
  }
}

:host([icon][disabled]) button bce-icon {
  color: bce-color-d(disable);
  background-color: bce-color(disable, 'icon');
}

// :host([icon]:not([design='text']):not([design='outline'])) button {
//   bce-icon {
//     left: auto;
//     right: 0;
//   }
//
//   span {
//     padding-left: 0;
//     padding-right: 36px;
//   }
// }

:host([design='outline']) button,
:host([design='text']) button {
  bce-icon {
    background-color: transparent;
  }

  &:after {
    display: none;
  }
}

:host([design='outline'][icon]) button,
:host([design='text'][icon]) button {
  span {
    padding-left: 22px;
  }

  &[data-icon-only] bce-icon {
    background-color: transparent;
    color: bce-color();
  }
}

:host([design='outline'][disabled]) button bce-icon,
:host([design='text'][disabled]) button bce-icon {
  background-color: transparent;
}

:host([design='text']) {
  border-radius: 4px;
}
