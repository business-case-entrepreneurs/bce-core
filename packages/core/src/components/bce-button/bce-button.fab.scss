@import './bce-button.scss';

:host {
  @include bce-color-inherit();
}

:host([icon]) {
  @include elevate(0, false);
  height: 40px;
  margin: 0 10px 10px 0;
  background-color: transparent;
  z-index: 6;

  button {
    bce-icon {
      @include elevate(2, false);
      left: auto;
      right: 0;
      top: 2px;
      width: 35px;
      height: 35px;
      background-color: bce-color();
      color: bce-color-t();
      border-radius: 50%;
    }

    span {
      @include elevate(2, false);
      margin: 0 30px 0 0;
      padding: 5px 16px;
      background-color: bce-color(light);
      color: bce-color-t(light);
      text-transform: none;
      border-radius: 34px;
    }

    &:before {
      display: none;
    }
  }
}

:host([data-inactive]) {
  display: none;
}

:host([data-inline]) {
  flex: 0 0 50%;
  margin: 0;

  button {
    padding: 0;

    span {
      display: block;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      margin: 0 40px 0 5px;
    }
  }

  @include media-medium {
    button {
      span {
        margin: 0 5px 0 40px;
      }

      bce-icon {
        left: 0;
        right: auto;
      }
    }
  }
}

:host([data-inline]:nth-child(odd)) button {
  span {
    margin: 0 5px 0 40px;
  }

  bce-icon {
    left: 0;
    right: auto;
  }
}
