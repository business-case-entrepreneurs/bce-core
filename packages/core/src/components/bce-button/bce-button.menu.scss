@import './bce-button.scss';

:host {
  @include bce-color-inherit();
  @include elevate(0);
  display: flex;
  margin: 0;

  button {
    text-align: left;
  }
}
