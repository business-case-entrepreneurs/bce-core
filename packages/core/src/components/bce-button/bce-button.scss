:host {
  @include bce-color-default(primary);
  @include box-sizing();
  @include elevate(2);
  position: relative;
  display: inline-block;
  height: 36px;
  margin: 0 2px;
  background-color: bce-color();
  color: bce-color-t();
  border-radius: 4px;
  font-size: 14px;
  font-weight: 500;
  letter-spacing: 0.1em;
  text-transform: uppercase;
  overflow: hidden;
  user-select: none;
  transition: box-shadow 0.28s cubic-bezier(0.4, 0, 0.2, 1);
  cursor: pointer;

  bce-ripple {
    background-color: bce-color-t();
  }

  button {
    width: 100%;
    height: 100%;
    padding: 0 12px;
    font-family: inherit;
    font-weight: inherit;
    font-size: inherit;
    letter-spacing: inherit;
    text-transform: inherit;
    background-color: transparent;
    color: inherit;
    border: none;
    appearance: none;
    cursor: inherit;

    &:before {
      content: '';
      position: absolute;
      left: 0;
      right: 0;
      top: 0;
      bottom: 0;
      pointer-events: none;
      background-color: bce-color-t();
      opacity: 0;
    }

    &:hover:before {
      opacity: 0.1;
    }

    &:focus {
      outline: none;

      &:before {
        opacity: 0.2;
      }
    }
  }
}

:host([block]) {
  width: 100%;
}

:host([disabled]) {
  @include elevate(0);
  color: bce-color-t(disable);
  background-color: bce-color(disable);
  opacity: 0.5;
  cursor: default;

  button:before {
    display: none;
  }
}

:host([focus]) {
  @include elevate(4);
  outline: none;
  -webkit-tap-highlight-color: rgba(255, 255, 255, 0);
  -webkit-tap-highlight-color: transparent;
}

:host([icon]) button {
  span {
    padding-left: 24px;
  }

  bce-icon {
    position: absolute;
    height: 100%;
    top: 0;
    left: 0;
    display: inline-flex;
    justify-content: center;
    align-items: center;
  }

  &.icon-only {
    padding: 0;

    bce-icon {
      position: relative;
    }

    span {
      display: none;
    }
  }
}

:host([small]) {
  height: 28px;

  button {
    padding: 0 8px;

    span {
      padding-left: 18px;
    }
  }
}

:host([design='outline']) {
  border: 1px solid bce-color-d(light);

  button {
    padding: 0 11px;
  }
}

:host([design='outline']),
:host([design='text']) {
  @include elevate(0);
  background-color: transparent;
  color: bce-color();

  bce-ripple {
    background-color: bce-color();
  }

  button:before {
    background-color: bce-color();
  }
}

:host([design='outline'][disabled]),
:host([design='text'][disabled]) {
  color: bce-color-t(disable);
}

:host([design='outline'][small]) button {
  padding: 0 7px;
}

:host([design='text'][icon][block]) > button {
  > span {
    padding-left: 0;
  }

  > bce-icon {
    top: 2px;
    position: relative;
  }
}

:host([upload]) {
  // Visually hide input
  > input[type='file'] {
    position: absolute !important;
    height: 1px;
    width: 1px;
    overflow: hidden;
    clip: rect(1px, 1px, 1px, 1px);
  }
}
