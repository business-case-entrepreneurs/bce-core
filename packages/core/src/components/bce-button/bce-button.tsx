import {
  Component,
  Element,
  Event,
  EventEmitter,
  getMode,
  h,
  Host,
  Prop,
  State
} from '@stencil/core';

import { ButtonDesign } from '../../models/button-design';
import { File } from '../../utils/file';
import { NativeEvent } from '../../utils/native-event';
import { ripple } from '../../utils/ripple';
import { UUID } from '../../utils/uuid';

@Component({
  tag: 'bce-button',
  styleUrls: {
    'bce-fab': 'bce-button.fab.scss',
    'bce-menu': 'bce-button.menu.scss',
    bucket: 'bce-button.bucket.scss',
    default: 'bce-button.scss'
  },
  shadow: true
})
export class Button {
  @Element()
  private el!: HTMLBceButtonElement;

  // #region Custom properties
  @Prop({ reflect: true })
  public color?: string;

  @Prop({ reflect: true })
  public design?: ButtonDesign;

  @Prop({ reflect: true })
  public icon?: string;

  @Prop({ reflect: true })
  public iconSpin?: boolean;

  @Prop({ reflect: true })
  public block?: boolean;

  @Prop({ reflect: true })
  public small?: boolean;

  @Prop({ reflect: true, attribute: 'focus' })
  public hasFocus?: boolean;

  @Prop({ reflect: true })
  public upload?: boolean;
  // #endregion

  // #region Forwarded to native button
  @Prop({ reflect: true })
  public disabled = false;

  @Prop({ reflect: true })
  public form?: string;

  @Prop({ reflect: false, attribute: 'formaction' })
  public formAction?: string;

  @Prop({ reflect: false, attribute: 'formenctype' })
  public formEnctype?:
    | 'application/x-www-form-urlencoded'
    | 'multipart/form-data'
    | 'text/plain';

  @Prop({ reflect: false, attribute: 'formmethod' })
  public formMethod?: 'get' | 'post';

  @Prop({ reflect: false, attribute: 'formtarget' })
  public formTarget?: string;

  @Prop({ reflect: true })
  public type: 'button' | 'reset' | 'submit' = 'button';
  // #endregion

  // #region Forwarded to native input[type='file']
  @Prop({ reflect: true })
  public accept?: string;

  @Prop({ reflect: true })
  public multiple = false;
  // #endregion

  @Event({ eventName: 'file' })
  private onFile!: EventEmitter<File[]>;

  @State()
  private slotEmpty = false;

  private handleBlur = () => {
    this.hasFocus = false;
  };

  private handleClick = (event: NativeEvent) => {
    if (this.upload) {
      // Trigger input[type='file'] click
      const input = this.el.shadowRoot!.querySelector('input');
      if (input) input.click();

      // Let input's click event be the only propagated click event
      event.cancelBubble = true;
      event.preventDefault();
    }
  };

  private handleFocus = () => {
    this.hasFocus = true;
  };

  private handleMouseDown = (event: MouseEvent) => {
    if (this.disabled) return;
    ripple(this.el.shadowRoot!.querySelector('button')!, event);
  };

  private handleSlotChange = () => {
    this.slotEmpty = !this.el.childNodes.length;
  };

  private handleUpload = (event: NativeEvent) => {
    const input = event.target as HTMLInputElement | null;
    if (!input || !input.files) return;

    // Extract required data and generate id
    const files = Array.from(input.files).map(file => {
      return new File(UUID.v4(), file.name, file);
    });

    // Dispatch custom event
    this.onFile.emit(files);
  };

  componentDidLoad() {
    const slot = this.el.shadowRoot!.querySelector('slot');
    if (slot) {
      slot.addEventListener('slotchange', this.handleSlotChange);
      this.handleSlotChange();
    }
  }

  renderIcon() {
    const mode = getMode(this);
    const size = this.el.clientHeight + 'px';
    const style =
      mode === 'bce-fab' ? undefined : { width: size, height: size };

    return (
      <bce-icon
        style={style}
        raw={this.icon}
        spin={this.iconSpin}
        fixed-width
      />
    );
  }

  render() {
    return (
      <Host
        onBlur={this.handleBlur}
        onFocus={this.handleFocus}
        onMouseDown={this.handleMouseDown}
      >
        <button
          class={{ 'icon-only': this.slotEmpty }}
          disabled={this.disabled}
          form={this.form}
          formaction={this.formAction}
          formenctype={this.formEnctype}
          formmethod={this.formMethod}
          formtarget={this.formTarget}
          type={this.type}
          onClick={this.handleClick}
        >
          {this.icon && this.renderIcon()}
          <span>
            <slot />
          </span>
        </button>

        {this.upload && (
          <input
            type="file"
            accept={this.accept}
            multiple={this.multiple}
            onChange={this.handleUpload}
            tabIndex={-1}
          />
        )}
      </Host>
    );
  }
}
