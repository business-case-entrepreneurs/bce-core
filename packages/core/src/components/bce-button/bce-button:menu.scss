@import './bce-button.scss';

:host {
  @include bce-color-inherit();
  @include elevate(0);
  display: flex;
  margin: 0;
}
