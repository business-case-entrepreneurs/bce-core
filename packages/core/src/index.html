<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0"
    />
    <meta name="theme-color" content="#080707" />
    <title>BCE Core</title>

    <link rel="stylesheet" href="/build/bce.css" />
  </head>
  <body>
    <bce-root id="app">
      <bce-status-bar color="dark"></bce-status-bar>
      <bce-header color="dark">
        <bce-logo logo-src="/logo/logo.png"></bce-logo>
        <bce-menu-button color="light"></bce-menu-button>
      </bce-header>

      <bce-menu color="dark" verticale right>
        <a href="#switch">Switch</a>
        <a href="#input">Input</a>
        <a href="#button">Button</a>
        <a href="#message">Message</a>
      </bce-menu>

      <main>
        <h3 id="dropdown">BCE Dropdown menu</h3>
        <div>
          <bce-dropdown-menu>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
          </bce-dropdown-menu>
        </div>
        <h3 id="slider">BCE Slider</h3>
        <div><bce-slider max="200"></bce-slider></div>
        <h3 id="switch">BCE Switch</h3>
        <div>
          <bce-switch value="true"></bce-switch>
          <bce-switch value="true" color="secondary"></bce-switch>
          <bce-switch value="true" color="tertiary"></bce-switch>
        </div>
        <div>
          <bce-switch value="true" color="success"></bce-switch>
          <bce-switch value="true" color="error"></bce-switch>
          <bce-switch value="true" color="warning"></bce-switch>
          <bce-switch value="true" color="info"></bce-switch>
        </div>
        <div>
          <bce-switch value="true" color="dark"></bce-switch>
          <bce-switch value="true" color="light"></bce-switch>
        </div>

        <h3 id="input">BCE Input</h3>
        <bce-input
          label="Username"
          value="my-username"
          info="What's your name?"
        ></bce-input>
        <bce-input
          type="password"
          label="Password"
          info="This should be a value only you know..."
        ></bce-input>
        <bce-input
          type="number"
          label="OTP code"
          info="The One Time Password code."
          disabled
        ></bce-input>

        <bce-input type="dropdown" label="Dropdown" info="testing..">
          <option value="A">Option A</option>
          <option value="B">Option B</option>
          <option value="C">Option C</option>
          <option value="D">Option D</option>
          <option value="E">Option E</option>
          <option value="F">Option F</option>
          <option value="G">Option G</option>
        </bce-input>

        <h3 id="button">BCE Button</h3>
        <h4>Contained</h4>
        <div>
          <bce-button>Primary</bce-button>
          <bce-button color="secondary">Secondary</bce-button>
          <bce-button color="tertiary">Tertiary</bce-button>
          <bce-button icon="square">Icon</bce-button>
          <bce-button disabled>Disabled</bce-button>
        </div>
        <br />
        <h4>Outline</h4>
        <div>
          <bce-button type="outline">Primary</bce-button>
          <bce-button type="outline" color="secondary">Secondary</bce-button>
          <bce-button type="outline" color="tertiary">Tertiary</bce-button>
          <bce-button type="outline" icon="square">Icon</bce-button>
          <bce-button type="outline" disabled>Disabled</bce-button>
        </div>
        <br />
        <h4>Text</h4>
        <div>
          <bce-button type="text">Primary</bce-button>
          <bce-button type="text" color="secondary">Secondary</bce-button>
          <bce-button type="text" color="tertiary">Tertiary</bce-button>
          <bce-button type="text" icon="square">Icon</bce-button>
          <bce-button type="text" disabled>Disabled</bce-button>
        </div>
        <h4>Block</h4>
        <div>
          <bce-button color="tertiary" block>Contained</bce-button>
          <bce-button icon="square" block>Contained + Icon</bce-button>
          <bce-button type="outline" color="tertiary" block>Outline</bce-button>
          <bce-button type="outline" icon="square" block>
            Outline + Icon
          </bce-button>
          <bce-button type="text" color="tertiary" block>Text</bce-button>
          <bce-button type="text" icon="square" block>Text + Icon</bce-button>
        </div>

        <h3>BCE message</h3>
        <bce-button id="message">Trigger</bce-button>

        <h3 id="dropdown">BCE Dropdown menu</h3>
        <div>
          <bce-dropdown-menu>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
          </bce-dropdown-menu>
        </div>

        <h3>Heading 3</h3>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipisicing elit.
          Reprehenderit assumenda omnis sint, tenetur explicabo optio quam natus
          perferendis. Cupiditate perspiciatis ea, assumenda libero enim
          praesentium atque repellendus voluptatem explicabo perferendis.
        </p>
        <p>
          Fugiat praesentium distinctio officiis expedita consequuntur vitae
          amet ipsa enim odit nesciunt facere alias aliquid cupiditate
          perferendis, nobis aut qui velit maiores recusandae sed? Facilis
          aliquam rem autem! Dolores, modi?
        </p>
        <p>
          Sunt rem praesentium exercitationem eveniet excepturi impedit delectus
          qui similique quaerat! Fugit obcaecati sint quaerat animi cumque a
          beatae? Error, facilis necessitatibus? Provident, perferendis
          doloribus? Delectus quibusdam fugiat nulla vel!
        </p>
        <p>
          Quis delectus iure perferendis exercitationem suscipit necessitatibus
          doloribus? In dolore voluptates temporibus odit sapiente ea provident
          rerum, officia et magni consequuntur esse est unde, cumque accusamus
          dignissimos? Ullam, sit quidem?
        </p>
        <p>
          Aliquam odio excepturi quis reiciendis, recusandae dolor at, quam
          ipsam sunt magni modi beatae libero maiores similique amet iusto minus
          aspernatur quisquam. Veritatis, ducimus illum. Fugiat iure dolores
          assumenda voluptatum!
        </p>
        <p>
          Aperiam beatae quo cupiditate hic recusandae ad laudantium
          voluptatibus fugiat odit, harum sunt officia sed eaque unde ipsam
          obcaecati non magni voluptatem adipisci quia dignissimos dolores modi.
          Optio, tempora mollitia.
        </p>

        <bce-fab color="dark">
          <bce-button>TODO</bce-button>
          <bce-button>animate</bce-button>
          <bce-button>floating</bce-button>
          <bce-button>action</bce-button>
          <bce-button>button</bce-button>
          <bce-button>(de)activation</bce-button>
        </bce-fab>
      </main>
    </bce-root>

    <script type="module" src="build/bce.esm.js"></script>
    <script nomodule src="build/bce.js"></script>
    <script>
      var root = document.getElementById('app');
      var message = document.getElementById('message');

      var i = 0;
      message.addEventListener('click', function() {
        root.message('Message ' + ++i);
      });
    </script>
  </body>
</html>
