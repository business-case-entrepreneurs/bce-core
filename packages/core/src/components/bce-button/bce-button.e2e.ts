import { newE2EPage, E2EPage } from '@stencil/core/testing';
// import { Button } from './bce-button';

describe('Diff test', () => {
  it('render something', async () => {
    const page: E2EPage = await newE2EPage();
    await page.setContent('<bce-button>Test</bce-button>');

    // To start comparing the visual result, you first must run page.compareScreenshot; This will capture a screenshot, and save the file to "/screenshot/images". You'll be able to check that into your repo to provide those results to your team. You can only have one of these commands per test.
    const results = await page.compareScreenshot();

    // Finally, we can test against the previous screenshots.
    // Test against hard pixels
    expect(results).toMatchScreenshot({ allowableMismatchedPixels: 100 });

    // Test against the percentage of changes. if 'allowableMismatchedRatio' is above 20% changed,
    expect(results).toMatchScreenshot({ allowableMismatchedRatio: 1 });

    // await page.compareScreenshot(
    //   'My Componment (...is beautiful. Look at it!)',
    //   { fullPage: false }
    // );
  });
});
