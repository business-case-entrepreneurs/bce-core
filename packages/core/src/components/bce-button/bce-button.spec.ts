import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { Button } from './bce-button';

describe('Testing bce-button', () => {
  let page: SpecPage;

  beforeEach(async () => {
    page = await newSpecPage({
      components: [Button],
      html: `<bce-button></bce-button>`
    });
  });

  it('render bce-button', async () => {
    expect(page.root).toMatchSnapshot();
  });
});
