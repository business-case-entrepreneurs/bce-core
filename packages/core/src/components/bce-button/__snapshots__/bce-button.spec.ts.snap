// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing bce-button render bce-button 1`] = `
<bce-button type="button">
  <mock:shadow-root>
    <button class="icon-only" type="button">
      <span>
        <slot></slot>
      </span>
    </button>
  </mock:shadow-root>
</bce-button>
`;
